docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,20 +2,228 @@
+ |3             |    | p1 |      |     |     |   |   | p2|A(i)+=VP(i)
+ |              |    | p1 |      | p2  |     |   |   |   |A(p2)-= P
+ |4             |    |    |   p1 |     |     |   | p2|   |A(i)+=VP(i)
+ |              |    |    |p1,p2 |     |     |   |   |   |A(p2)-= P
+
+The actual implementation does not create this queue, it only has to store the current accumulated proposer priority. 
+
+There are cases where the proposer selection is executed multiple times.
+(find a good example here)
+So we modify slightly the algorithm to handle this case. Note that times indicates the number of executions of the simpleProposalSelection() 
+
+    def singleProposalSelection (vset):
+        for each validator i in vset:
+            A(i) += VP(i)
+        prop = max(A)
+        A(prop) -= P
+    
+    def proposalSelection (vset, times):
+        for in in range(times):
+            singleProposalSelection(vset)
+
+### Validator Set Changes
+At each block height the validator set may change. Some of the changes have implications on the proposer selection.
+
+#### Voting Power Change
+Consider again the earlier example and assume that the voting power of p1 is changed to 6:
+
+Validator | p1 | p2 
+----------|--- | ---
+VP        | 4  | 3
+
+Let's also assume that the last round R before this change the proposer priorites were as shown in first row (R-1 end). As it can be seen, the procedure can continue with round 1 at next height without changes as before. 
+
+|Priority Round| -2 | -1 | 0    | 1   | 2   | 3 | 4 | 5 | Comment
+|--------------| ---|--- |------|---  |---  |---|---|---|--------
+| R end        |    | p2 |      |  p1 |     |   |   |   |update VP(p1)
+| 1            |    |    |      |     | p2  |   |   | p1|A(i)+=VP(i)
+|              | p1 |    |      |     | p2  |   |   |   |A(p1)-= P
+
+#### Validator Removal
+Consider a new example with set.
+
+Validator | p1 | p2 | p3 |
+--------- |--- |--- |--- |
+VP        | 1  | 2  | 3  |
+
+Let's assume that the last round was R and the proposer priorities were as shown in first row (R end) with their sum being 0. At this point p2 is removed and round 1 of next height runs. At the end of the round (penultimate row) the sum of priorities is -2 (minus the priority of the removed process). 
+ 
+
+|Priority Round |-3 | -2 | -1 | 0  | 1   | 2   | 3 | 4 | 5 | Comment
+|---------------|-- | ---|--- |--- |---  |---  |---|---|---|--------
+| R end         |p3 |    |    |    | p1  | p2  |   |   |   |remove p2
+| 1             |   |    |    | p3 |     | p1  |   |   |   |A(i)+=VP(i)
+|               |   | p1 |    | p3 |     |     |   |   |   |A(p1)-= P
+| *new step*    |   |    | p1 |    | p3  |     |   |   |   |A(i) -= avg
+
+The procedure could continue without modifications. However, it is possible that after a sufficiently large number of modifications in validator set, the priority values would group towards maximum or minimum allowed values.
```

Comment:
Can we clarify why this is a problem - ie. it will lead to overflow issues if we're using standard integers. Is there some other reason?

Reply:
done